lib/sensors/src/optical_flow.rs

```diff
@@ -0,0 +1,365 @@
+use defmt::*;
+use embassy_time::{Duration, Instant, Timer};
+use hyped_gpio::HypedGpioOutputPin;
+use hyped_spi::{HypedSpi, SpiError};
+
+/// Optical flow implements the logic to interact with the PMW3901MB-TXQT: Optical Motion Tracking Chip
+///
+/// This implementation is directly coming from https://github.com/pimoroni/pmw3901-python/blob/main/pmw3901/__init__.py
+/// Data Sheet: https://www.codico.com/de/mpattachment/file/download/id/952/
+pub struct OpticalFlow<'a, T: HypedSpi + 'a, C: HypedGpioOutputPin> {
+    spi: &'a mut T,
+    cs: C,
+}
+
+impl<'a, T: HypedSpi, C: HypedGpioOutputPin> OpticalFlow<'a, T, C> {
+    pub async fn new(spi: &'a mut T, cs: C) -> Result<Self, OpticalFlowError> {
+        let mut optical_flow = Self { spi, cs };
+
+        optical_flow.cs.set_low();
+        Timer::after(Duration::from_millis(5)).await;
+        optical_flow.cs.set_high();
+
+        Timer::after(Duration::from_millis(2)).await;
+        optical_flow.write(REG_POWER_UP_RESET, POWER_UP_RESET_INSTR)?;
+
+        for offset in 0..NUM_UNIQUE_DATA_VALUES {
+            optical_flow.read(REG_DATA_READY + offset)?;
+        }
+
+        optical_flow.secret_sauce().await?;
+        debug!("Secret sauce done");
+
+        let (product_id, revision_id) = optical_flow.get_id()?;
+        if product_id != PMW3901_PRODUCT_ID {
+            warn!("Invalid product id: {}", product_id);
+        }
+
+        if !VALID_PMW3901_REVISIONS.contains(&revision_id) {
+            warn!("Invalid revision id: {}", revision_id);
+        }
+
+        Ok(optical_flow)
+    }
+
+    /// Gets the product ID and revision ID to verify the sensor is connected
+    fn get_id(&mut self) -> Result<(u8, u8), OpticalFlowError> {
```

Comment:
why not just inline this? packing stuff into then straight back out of pairs feels bad (on vibes only so ignore if you want)